### PR TITLE
Android build standup

### PR DESCRIPTION
**Pull request by a1studmuffin**
Hey team, I've spent the last few weeks burning the midnight oil standing up the Android build in mainline. With this PR you can now compile and run Cataclysm DDA on an Android device. See my additions to COMPILING.md for an overview and instructions on how to build an Android APK under Linux.

There is a new folder in the root of the repository `android/` which contains the Android project files.

The rest of the changes are to game C++ files in `src/` (as well as an Android.mk file). Almost all of the C++ changes are guarded for `__ANDROID__` so they won't affect other platforms, with the exception of one or two minor additions.

I'd appreciate if anyone wants to skim over the changes and offer feedback. If you're attempting to get this running under Windows with Android Studio (I've got it building under Ubuntu and Windows here), ensure your git supports creation of symlinks.

Most of the C++ changes to random game files are just setting up input_contexts specifically for Android, often with manually registered buttons + descriptions to match the existing UI code. This is so that the Android quick shortcuts along the bottom of the touchscreen work even in cases where the game's UI code hasn't been refactored to use an input_context yet. I considered rewriting each of these UI screens to use input_contexts for all platforms, but that's a reasonably big job and I was hoping we could tackle it later on.

There's also a big hunk of Android-specific code sitting in sdltiles.cpp. Let me know if this is unpalatable and I can look at merging some of it into a separate C++ file containing Android-specific code, perhaps android.cpp/h.

Things that are still outstanding (but can happen after this PR):
- Crash reporting (would be very useful to improve game reliability across devices)
- Release APK signing with a keystore (though this shouldn't live in the public repository anyway)
- Gradle project build steps for auto-generation of version.h and catabindings.cpp
- SDL software renderer is still unreliable on a lot of devices
- No Android-specific welcome screen. In my original port (https://github.com/a1studmuffin/Cataclysm-DDA-Android) I had a welcome screen explaining the Android touch controls. I've removed this to minimize C++ changes, but we may want to add something like this back in.
- Custom data folder support (allowing more flexible user-control over save game locations, syncing with desktop versions via Dropbox/Google Drive synced folders on their device etc.)

Completed:
- APKs can be thinned out as they currently contain lang files other than lang/mo (which aren't needed and take up about 25% of APK size) (175fc2d25d03c46cca95c5043f7217c93e2740e9)

### CODE REVIEW

**Comment by a1studmuffin** [comment]
The latter. With this change you can manually compile and run Cataclysm DDA
on an Android device. Once Android support has been integrated into
mainline, I'll work on getting nightly experimental Android builds up and
running alongside the other builds.

**Comment by ZhilkinSerg** [comment]
Sorry for waiting for too long in the PR queue. I have Linux VM ready now, so I will try to compile this next week.

**Comment by a1studmuffin** [comment]
No problem at all, it's a decent chunk of code so I figured it'd be a while before someone had the time to check it out - thanks for having a look! I'll continue keeping the PR updated to latest to ensure there's no merge conflicts. Let me know if you run into any issues.

File: COMPILING.md [hunk 1]
```diff
@@ -282,6 +283,92 @@ To build full curses version with localizations and lua enabled:
...
+Install Linux dependencies. For a desktop Ubuntu installation:
+
+    sudo apt-get install lua5.2 openjdk-8-jdk-headless
+
+Install Android SDK and NDK:
+
+    wget https://dl.google.com/android/repository/sdk-tools-linux-4333796.zip
+    unzip sdk-tools-linux-4333796.zip -d ~/android-sdk
+    rm sdk-tools-linux-4333796.zip
+    ~/android-sdk/tools/bin/sdkmanager --update
```
Comment (ZhilkinSerg) [inline COMPILING.md]:
```
zhilkin@ubuntu:~$ ~/android-sdk/tools/bin/sdkmanager --update
Warning: File /home/zhilkin/.android/repositories.cfg could not be loaded.      
[=======================================] 100% Computing updates...             
```

**Comment by ZhilkinSerg** [comment]
I guess it is working, but my VM hung up on 53% :(

image

File: COMPILING.md [hunk 2]
```diff
@@ -282,6 +283,92 @@ To build full curses version with localizations and lua enabled:
...
+Export Android environment variables (you can add these to the end of `~/.bashrc`):
+
+    export ANDROID_SDK_ROOT=~/android-sdk
+    export ANDROID_HOME=~/android-sdk
+    export ANDROID_NDK_ROOT=~/android-sdk/ndk-bundle
+    export PATH=$PATH:$ANDROID_SDK_ROOT/platform-tools
+    export PATH=$PATH:$ANDROID_SDK_ROOT/tools
+    export PATH=$PATH:$ANDROID_NDK_ROOT
+
+Install custom Android dependencies. From the root folder of the CDDA repository run:
```
Comment (ZhilkinSerg) [inline COMPILING.md]: I believe we should switch to `Cataclysm-DDA` folder before cloning.

**Comment by ZhilkinSerg** [comment]
Should I make Linux version with languages first?

image

**Comment by ZhilkinSerg** [comment]
Didn't test it on the phone yet, but apk was built:

image

File: COMPILING.md (hunk 1 above)
Comment (a1studmuffin) [reply COMPILING.md]: I believe this warning is "normal" for this Android toolchain version. It appears if the file it's looking for doesn't exist locally, but doesn't impact anything after that: https://askubuntu.com/questions/885658/android-sdk-repositories-cfg-could-not-be-loaded

File: COMPILING.md (hunk 2 above)
Comment (a1studmuffin) [reply COMPILING.md]: Ah thanks, that's what I meant to say with "From the root folder of the CDDA repository run:" but it's probably clearer to say "Switch to the `Cataclysm-DDA` folder then run:". I'll update that.

**Comment by Dyrewulfe** [comment]
@a1studmuffin 

I've built a branch off of this PR that uses precompiled dependencies(easier end user builds, and likely easier for the CI machinery to deal with), and adds some per build type flags for optimization. I'll PR it at your repo once I've finished testing everything out, if you're interested?(I felt I should say something early, before you rebase again.)

File: COMPILING.md [hunk 3]
```diff
@@ -282,6 +283,92 @@ To build full curses version with localizations and lua enabled:
...
+
+    ./gradlew assembleDebug
+
+This creates a debug APK in `./android/app/build/outputs/apk/` ready to be installed on your device.
+
+To build a debug APK and immediately deploy to your connected device over adb run:
+
+    ./gradlew installDebug
+
+To build a signed release APK (ie. one that can be installed on a device), [build an unsigned release APK and sign it manually](https://developer.android.com/studio/publish/app-signing#signing-manually).
```
Comment (ZhilkinSerg) [inline COMPILING.md]: Might be useful to replicate commands from instruction from the link to create key, align apk and sign it.

Also had to install additional utilities to align apk zip and sign it:

```
sudo apt-get install zipalign
sudo apt-get install apksigner
```

**Comment by ZhilkinSerg** [comment]
Successfully created, installed and run Release version. Initially had some troubles with signing apk, but read the instructions and everything went smooth.

**Comment by a1studmuffin** [comment]
Woohoo! Thanks to everyone who tested this out, much appreciated.

**Pull request merged**